### ISSUE TITLE
no-extra-parens jsxIgnore multi-line equivalent?

### ISSUE THREAD

**Issue by polarathene** [yannickcr/eslint-plugin-react#2624]
The airbnb docs have a comment for the `no-extra-parens` rule, citing `jsxIgnore` should be ignored there, and [`react/jsx-wrap-multlines`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-wrap-multilines.md) should be used for handling JSX. However that only seems to cover multi-line JSX as the rule name implies.

`no-extra-parens` can allow for parentheses around multi-line JSX, while also detecting and avoiding it for single line JSX, as per the airbnb guide advice examples. Prettier likewise formats JSX in the same manner with parens for multi-line, and stripping them if used in single line JSX return.

Is this a rule that should be catered for by this plugin? Or should `no-extra-parens` be used?

Related issue on the [airbnb JS style guide repo](https://github.com/airbnb/javascript/issues/2209).

**Comment by ljharb** [comment]
It seems like the gap you're referring to is that the airbnb guide recommends single-line jsx being wrapped in parens, but there's no linter rule in this plugin to cover that.

The intention of the airbnb guide is that the developer should be able to *choose* whether to use, or omit, parens around single-line jsx. Prettier destroys that choice by always stripping them. Do you have a use case to always include them? (note that the airbnb guide does not mandate this)

**Comment by polarathene** [comment]
> It seems like the gap you're referring to is that the airbnb guide recommends single-line jsx being wrapped in parens, but there's no linter rule in this plugin to cover that.

[The opposite of that is recommended by airbnb](https://airbnb.io/javascript/react/#parentheses)(no parentheses):

```jsx
// good, when single line
render() {
  const body = <div>hello</div>;
  return <MyComponent>{body}</MyComponent>;
}
```

This is the same as what Prettier does. My issue is that the general JS guide by airbnb provides a eslint config that defers the JSX option of `no-extra-parens` rule to this react plugin instead, however the plugin cannot reproduce the same behavior and autofix that the `no-extra-parens` rule is capable of.

> or omit, parens around single-line jsx.

This is not an option via this plugin, it either needs to support that or the config for `no-extra-parens` rule in the main airbnb JS config should not advise this plugin over that rules option for handling JSX.

> Do you have a use case to always include them? (note that the airbnb guide does not mandate this)

I exclude them from single line if they're not necessary with JSX, but error for multi-line, both can be auto fixed with eslint. I don't use Prettier personally as it often formats code in an undesirable way that cannot be avoided without noisy disable comments.

It may not be mandated on how to handle parentheses with JSX or not, but the link to the guide references what airbnb considers good style for both cases.

---

Either this plugin should support the equivalent of:

```js
"no-extra-parens": ["error", "all", { ignoreJSX: "multi-line" }],
```

Or [this comment](https://github.com/airbnb/javascript/blob/06b3ab11d9a443ff46f052dd00375e271e5146e6/packages/eslint-config-airbnb-base/rules/errors.js#L71)  should be modified or removed:

```js
ignoreJSX: 'all', // delegate to eslint-plugin-react 
```

**Comment by ljharb** [comment]
no-parens "good, when single line" does not imply that with parens would be bad.

**Comment by ljharb** [comment]
If the airbnb config modified its `no-extra-parens` config to only ignore multiline jsx, it would be forbidding parens on single-line jsx - which is GOOD and ALLOWED, just not forced by the guide. This is because the guide intentionally leaves it up to the individual developer to decide if single-line JSX should have parens or not - the tooling should not get an opinion on that.

**Comment by polarathene** [comment]
That's fine, but it should have a comment about that, rather than ignoring the rule for JSX with a comment that delegates responsibility to this plugin when this plugin cannot handle the single line behaviour in the same manner, only multi-line.

The comment could instead advise to use the related rule provided by the react plugin when only interested in multi-line linting, and ignoring consistency in single-line JSX handling. 

At the very least, it should be clear which rule in the react plugin it's deferring configuration to, and that rule should be clear about not supporting single-line(it roughly is via the rule name, but as a user it's not clear how to handle the single-line case while trusting the advice from the airbnb rule comment to ignore JSX for `no-extra-parens`.

**Comment by ljharb** [comment]
I'm not really sure what you're looking for here. The airbnb config's choice has nothing to do with this repo. The comments inside the airbnb config are for its maintainers (ie, me) and not for consumers. Separately, even if eslint-plugin-react could handle the single-line form, the airbnb config *would not enable that* - the desired behavior is that single-line jsx parens are not linted whatsoever, in either direction.

If there's something actionable in this repo, I'm happy to reopen - otherwise I think this can be closed.

**Comment by polarathene** [comment]
> The comments inside the airbnb config are for its maintainers (ie, me) and not for consumers.

Is that mentioned anywhere? As a consumer it seems a bit misleading to mention the functionality is handled by the react plugin, when it's not 1:1 parity in functionality.

> the airbnb config would not enable that - the desired behavior is that single-line jsx parens are not linted whatsoever, in either direction.

Ok.. no problem I can just use the non-react plugin rule as mentioned without deferring to the plugin to get the consistent single line style behavior. 

Probably should remove the "good" example that was linked earlier from JSX no-parens example, as that seems rather misleading too (as in if the other style isn't "bad", it's unclear how that is "good"), although from what I understand, there is some line where something is just advised as good style, but no enforcement/support from config, even when that is possible...? (I think that's not uncommon for consumers)

Cheers.

**Comment by ljharb** [comment]
It doesn’t say that - it says “delegated”.
